opm/common/OpmLog/MessageLimiter.hpp

```diff
@@ -46,62 +51,133 @@ namespace Opm
+        {
+            return tag_limit_;
+        }
+
+        /// The category message limits.
+        const std::map<int64_t, int>& categoryMessageLimits() const
+        {
+            return category_limits_;
+        }
+
+        /// The category message counts.
+        const std::map<int64_t, int>& categoryMessageCounts() const
+        {
+            return category_counts_;
+        }
+
+        /// Used for handleMessageLimits() return type (see that
         /// function).
         enum class Response
         {
-            PrintMessage, JustOverLimit, OverLimit
+            PrintMessage, JustOverTagLimit, JustOverCategoryLimit, OverTagLimit, OverCategoryLimit
         };
 
-        /// If a tag is empty, there is no message limit or for that
-        /// tag (count <= limit), respond PrintMessage.
-        /// If (count == limit + 1), respond JustOverLimit.
-        /// If (count > limit + 1), respond OverLimit.
-        Response handleMessageTag(const std::string& tag)
+        /// If (tag count == tag limit + 1) for the passed tag, respond JustOverTagLimit.
+        /// If (tag count > tag limit + 1), respond OverTagLimit.
+        /// If a tag is empty, there is no tag message limit or for that tag
+        /// (tag count <= tag limit), consider the category limits:
+        /// If (category count == category limit + 1) for the passed messageMask, respond JustOverCategoryLimit.
+        /// If (category count > category limit + 1), respond OverCategoryLimit.
+        /// If (category count <= category limit), or there is no limit for that category, respond PrintMessage.
+        Response handleMessageLimits(const std::string& tag, const int64_t messageMask)
         {
-            if (tag.empty() || message_limit_ == NoLimit) {
-                return Response::PrintMessage;
-            } else {
+            Response res = Response::PrintMessage;
+
+            // Deal with tag limits.
+            if (!tag.empty() && tag_limit_ != NoLimit) {
                 // See if tag already encountered.
                 auto it = tag_counts_.find(tag);
                 if (it != tag_counts_.end()) {
                     // Already encountered this tag. Increment its count.
                     const int count = ++it->second;
-                    return countBasedResponse(count);
+                    res = countBasedResponseTag(count);
                 } else {
                     // First encounter of this tag. Insert 1.
                     tag_counts_.insert({tag, 1});
-                    return countBasedResponse(1);
+                    res = countBasedResponseTag(1);
+                }
+            }
+
+            // If tag count reached the limit, the message is not counted
+            // towards the category limits.
+            if (res == Response::PrintMessage) {
+                // We are *not* above the tag limit, consider category limit.
+                const int count = ++category_counts_[messageMask];
```

Comment:
Actually, it will not be counted by the CounterLog, since it (or any other log backend) will never see the messages that are over the tag limit! If we want to have the "all messages including those over the tag limit" counts separate from the "printed messages" counts, it must be maintained in the MessageLimiter, too. Then I think the CounterLog can be implemented simply by asking the MessageLimiter for those numbers.

However the current implementation *really* ignores messages above the tag limits, meaning they won't even be counted towards the total (to avoid such messages filling up the category limit).